calendar-bundle/src/EventListener/BreadcrumbListener.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CalendarBundle\EventListener;
+
+use Contao\CalendarEventsModel;
+use Contao\CalendarModel;
+use Contao\Config;
+use Contao\CoreBundle\Framework\Adapter;
+use Contao\CoreBundle\Framework\ContaoFrameworkInterface;
+use Contao\Events;
+use Contao\Input;
+use Contao\PageModel;
+use Contao\StringUtil;
+
+class BreadcrumbListener
+{
+    /**
+     * @var ContaoFrameworkInterface
+     */
+    private $framework;
+
+    public function __construct(ContaoFrameworkInterface $framework)
+    {
+        $this->framework = $framework;
+    }
+
+    public function onGenerateBreadcrumb(array $items): array
+    {
+        $eventAlias = $this->getEventAlias();
+        if (!$eventAlias) {
+            return $items;
+        }
+
+        $calendar = $this->getCalenderModel();
+        if (!$calendar) {
+            return $items;
+        }
+
+        $eventModel = $this->getEventModel($eventAlias, $calendar);
+        if (!$eventModel) {
+            return $items;
+        }
+
+        if ($GLOBALS['objPage']->requireItem) {
+            return $this->overrideActiveBreadcrumbItem($items, $eventModel);
+        }
+
+        return $this->addBreadcrumbItem($items, $eventModel);
+    }
+
+    private function getEventAlias(): ?string
+    {
+        if (!isset($GLOBALS['objPage'])) {
+            return null;
+        }
+
+        /** @var Adapter|Input $inputAdapter */
+        $inputAdapter = $this->framework->getAdapter(Input::class);
+        /** @var Adapter|Config $configAdapter */
+        $configAdapter = $this->framework->getAdapter(Config::class);
+
+        if ($configAdapter->get('useAutoItem')) {
+            return $inputAdapter->get('auto_item');
```

Comment:
This is already done [here](https://github.com/contao/contao/blob/master/calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php#L59-L62) and does not need to be done again.

Reply:
Only, if the news reader is loaded **before** the breadcrumb module. I don't want to limit it to this precondition.

Reply:
Ah, no, this assumption is wrong. Only the `EventReader` class can be sure that `auto_item` is supposed to be mapped to `events`. Right now, the listener adds an event breadcrumb item as soon as **any** auto_item parameter is present, without even knowing if the auto_item is related to an event!

Only the `EventReader` can assume: "If I am called, there should be an `events` parameter. If there is no `events` parameter but I am still called, I should probably look for an `auto_item` parameter."

Therefore the `BreadcrumbListener` must only check the `events` parameter – even if this means that the feature will only work if the breadcrumb menu is loaded after the event reader.

Reply:
Yes, I know that this is a limitation. However, it depends on the `jumpTo` url of the news/event/faq. Afaik it's not possible to multiple reader modules on a page. That's why the current solution makes sense for me.

Reply:
> Afaik it's not possible to multiple reader modules on a page.

Of course it is. You can have a news reader and an event reader on the same page and trigger one or another with the `items` or `events` parameter.

Reply:
> Of course it is. You can have a news reader and an event reader on the same page and trigger one or another with the items or events parameter.

Ah, I see. I was refering about using `auto_item` didn't thought about using the proper parameters.

Reply:
> Ah, no, this assumption is wrong. Only the EventReader class can be sure that auto_item is supposed to be mapped to events. Right now, the listener adds an event breadcrumb item as soon as any auto_item parameter is present, without even knowing if the auto_item is related to an event!

I have thought about it again and I don't see what's the problem here.

** How the reader modules work **

1. The reader modules loads a model regarding the `items` (faq, news) or `events` (events) parameter. For simplicity I talk only about the `items` parameter later on.
2. If the `useAutoItem` is defined the `auto_item` is mapped to the `items` parameter in the reader module.
3. If the reader module has a defined `items` parameter and does not find any published model, it will throw an `PageNotFoundException` exception.
4. If no `items` parameter is defined, an empty string is returned.
5. If two reader modules of different types are on the same page, it would only work if the event reader is combined with news or faq. Otherwise a module would trigger the `PageNotFoundException` exception.

** How this PR would work **

1. The breadcrumb listeners checks if the current page is defined as `jumpTo` page in the news archive / calendar / faq category setting
2. If this is true, it detects the `items` parameter regarding the `auto_item` setting
3. It try to find the model
4. If an model exists it would append or override the breadcrumb item regarding `requireItem` setting in the current page.

> Right now, the listener adds an event breadcrumb item as soon as any auto_item parameter is present, without even knowing if the auto_item is related to an event!

There are only two scenarios where this would become true:

1. `useAutoItem` is deactivated and news and faq reader is on the same page. 
2. `useAutoItem` is activated and two reader modules are on the same page.

Both scenarios have the issue: As soon as any `auto_item` is not found, a `PageNotFoundException` exception would be thrown. Only if they match the mentioned issue would occur. 

Is it a real project issue? Who uses two reader modules on the same page ensuring identical aliases for each published news/event/faq presented on the page?

Reply:
> Both scenarios have the issue: As soon as any `auto_item` is not found, a `PageNotFoundException` exception would be thrown.

The first scenario is perfectly valid, as both readers return an empty string if the `items` or `events` parameter is not set. Only the second scenario would always throw an exception.

Maybe we can solve this problem by some kind of "lazy module loading"? Right now, we process each module and immediately add its output in the `PageRegular::prepare()` method. However, we could add placeholders and defer the processing of certain modules.

Reply:
> The first scenario is perfectly valid, as both readers return an empty string if the items or events parameter is not set. Only the second scenario would always throw an exception.

No, they throw the `PageNotFoundException` exception:

https://github.com/contao/contao/blob/1caf237e78691ed12a823efd8a62f66ae0b9fffd/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php#L90-L97

https://github.com/contao/contao/blob/1caf237e78691ed12a823efd8a62f66ae0b9fffd/calendar-bundle/src/Resources/contao/modules/ModuleEventReader.php#L92-L99